## Gather

In [45]:
import pandas as pd

In [46]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [47]:
patients

patient_id assigned_sex  given_name       surname  \
0             1       female         Zoe       Wellish   
1             2       female      Pamela          Hill   
2             3         male         Jae        Debord   
3             4         male        Liêm          Phan   
4             5         male         Tim       Neudorf   
5             6         male      Rafael         Costa   
6             7       female        Mary         Adams   
7             8       female      Xiuxiu         Chang   
8             9         male       Dsvid    Gustafsson   
9            10       female      Sophie       Cabrera   
10           11       female       Sandy    Gunnarsson   
11           12         male   Abdul-Nur           Isa   
12           13         male  Omeokachie      Ibeamaka   
13           14       female    Anenechi         Chidi   
14           15       female        Asia       Woźniak   
15           16         male       Søren          Lund   
16           17       female         Tám          Liễu   
17           18       female     Roxanne     Andreyeva   
18           19         male     William         Oates   
19           20         male         Zak         Kelly   
20           21       female       Sofia       Karlsen   
21           22         male      Samúel  Guðbrandsson   
22           23         male      Manchu            Su   
23           24         male       Lovre         Galić   
24           25         male       Jakob      Jakobsen   
25           26         male      Gregor          Bole   
26           27       female        Ella          Lund   
27           28         male      Joseph        Tucker   
28           29         male      Robert          Wolf   
29           30         male        Jake      Jakobsen   
..          ...          ...         ...           ...   
473         474       female        Kate     Wilkinson   
474         475       female   Esperanza      Labrosse   
475         476         male       Malik       Vaneker   
476         477       female       Berta    Napolitani   
477         478         male     Juliusz      Majewski   
478         479       female      Edelma   Villalpando   
479         480         male        Tapa   Arsanukayev   
480         481         male      Nasser       Mansour   
481         482         male     Michael    Kristensen   
482         483         male       Diogo         Souza   
483         484       female       Angel         Grant   
484         485         male     Placido       Udinesi   
485         486         male      Trifon      Izmailov   
486         487         male      Samuel          Blix   
487         488         male        Ivar       Löfgren   
488         489         male        Mika    Martinsson   
489         490       female     Jasmine         Sykes   
490         491         male     Jackson       Addison   
491         492       female     Vanessa      Ferguson   
492         493         male       Poldi           Tar   
493         494       female         Fen          Chin   
494         495       female      Sirkka     Piirainen   
495         496         male      Hajime       Tsukada   
496         497         male   Alexander        Hueber   
497         498         male    Masataka      Murakami   
498         499         male     Mustafa     Lindström   
499         500         male       Ruman       Bisliev   
500         501       female       Jinke     de Keizer   
501         502       female     Chidalu   Onyekaozulu   
502         503         male         Pat       Gersten   

                        address               city       state  zip_code  \
0          576 Brown Bear Drive  Rancho California  California   92390.0   
1     2370 University Hill Road          Armstrong    Illinois   61812.0   
2         1493 Poling Farm Road               York    Nebraska   68467.0   
3           2335 Webster Street         Woodbridge          NJ    7095.0   
4          1428 Turkey Pen L

In [48]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
250         chen           yao          -  56u - 57u         7.90       7.51   
251        aksel   vestergaard          -  42u - 38u         9.62       9.29   
252        ellen         luman          -  40u - 39u         9.27       8.77   
253       albino     schiavone  35u - 43u          -         7.56       7.15   
254         jose         combs          -  39u - 36u         7.89       7.42   
255       jia li          teng  48u - 54u          -         7.66       7.32   
256        ilija        horvat  42u - 50u          -         7.77       7.38   
257     mathilde      nørgaard          -  27u - 28u         8.50       8.10   
258       csilla      herczegh          -  43u - 46u         7.71       7.27   
259      aaliyah          rice          -  31u - 31u         7.64       7.33   
260        david      beauvais          -  26u - 23u         7.87       7.47   
261     caroline        shuler          -  50u - 54u         7.63       7.27   
262         alex      crawford  51u - 62u          -         7.69       7.30   
263      rebecca      jephcott  53u - 63u          -         7.96       7.57   
264   chukwumoge     ogochukwu          -  41u - 39u         7.95       7.56   
265       fearne      mcgregor          -  27u - 29u         7.83       7.48   
266       ursula         freud  42u - 54u          -         7.75       7.46   
267         leon        scholz          -  38u - 32u         7.72       7.29   
268   

In [49]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [50]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [51]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [52]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [53]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [54]:
patients.columns

Index(['patient_id', 'assigned_sex', 'given_name', 'surname', 'address',
       'city', 'state', 'zip_code', 'country', 'contact', 'birthdate',
       'weight', 'height', 'bmi'],
      dtype='object')

In [55]:
treatments.columns

Index(['given_name', 'surname', 'auralin', 'novodra', 'hba1c_start',
       'hba1c_end', 'hba1c_change'],
      dtype='object')

In [56]:
adverse_reactions.columns

Index(['given_name', 'surname', 'adverse_reaction'], dtype='object')

In [57]:
all_columns

0           patient_id
1         assigned_sex
2           given_name
3              surname
4              address
5                 city
6                state
7             zip_code
8              country
9              contact
10           birthdate
11              weight
12              height
13                 bmi
14          given_name
15             surname
16             auralin
17             novodra
18         hba1c_start
19           hba1c_end
20        hba1c_change
21          given_name
22             surname
23    adverse_reaction
dtype: object

In [58]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [59]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [60]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [61]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [62]:
patients.sample(5)

patient_id assigned_sex given_name   surname                   address  \
106         107         male  Inunnguaq  Heilmann       4110 Capitol Avenue   
195         196       female     Lamara  Dratchev    3731 Swick Hill Street   
82           83       female       Châu     Thạch         4814 Layman Court   
84           85       female      Nương        Vũ       465 Southern Street   
455         456       female        Hui    Ch'eng  3423 Burning Memory Lane   

            city state  zip_code        country  \
106   New Castle    IN   47362.0  United States   
195  New Orleans    LA   70113.0  United States   
82       Atlanta    GA   30346.0  United States   
84      New York    NY   10001.0  United States   
455      Yardley    PA   19067.0  United States   

                                        contact   birthdate  weight  height  \
106   InunnguaqHeilmann@rhyta.com1 765 599 8847    5/6/1976   219.9      67   
195      985-253-7891LamaraDratchev@teleworm.us  10/20/1954   181.3      67   
82   678-350-2390ThachThiKieuChau@superrito.com    3/3/1999   171.4      60   
84           VuCamNuong@fleckens.hu516-720-5094    2/1/1981   138.2      63   
455              215-321-9611HuiCheng@rhyta.com    5/3/1982   203.1      69   

      bmi  
106  34.4  
195  28.4  
82   33.5  
84   24.5  
455  30.0

In [63]:
patients.surname.value_counts()

Doe             6
Jakobsen        3
Taylor          3
Ogochukwu       2
Lund            2
Grímsdóttir     2
Woźniak         2
Tạ              2
Liễu            2
Souza           2
Berg            2
Correia         2
Johnson         2
Gersten         2
Cabrera         2
Bùi             2
Schiavone       2
Nilsen          2
Kowalczyk       2
Hueber          2
Silva           2
Cindrić         2
Collins         2
Aranda          2
Dratchev        2
Kadyrov         2
Tucker          2
Batukayev       2
Lương           2
Parker          2
               ..
Horikawa        1
Morita          1
Onyekaozulu     1
Arsanukayev     1
Ekwueme         1
Hrafnsdóttir    1
Scheltens       1
Okoli           1
Dvořák         1
Kishiev         1
Woodward        1
Eldarkhanov     1
Milne           1
Olofsson        1
Stanić          1
Gaber           1
Hansen          1
Tsukada         1
Dandonneau      1
Labrosse        1
Hunter          1
Schneider       1
Jindrová        1
Carvalho        1
Zimmerman 

In [64]:
patients.address.value_counts()

123 Main Street             6
2476 Fulton Street          2
2778 North Avenue           2
648 Old Dear Lane           2
3414 Franklin Avenue        1
4500 Myra Street            1
3161 Fantages Way           1
2967 Prudence Street        1
4839 North Avenue           1
2866 Myra Street            1
4853 University Drive       1
3303 Anmoore Road           1
69 Mattson Street           1
4649 Joanne Lane            1
4792 Maud Street            1
206 Eagle Lane              1
2881 Harter Street          1
2438 Shady Pines Drive      1
1231 Grim Avenue            1
4093 Smith Street           1
1934 August Lane            1
2531 Eastland Avenue        1
4291 Patton Lane            1
3145 Sheila Lane            1
783 Callison Lane           1
2322 Grim Avenue            1
2972 Hillview Street        1
4148 Callison Lane          1
2146 Willow Greene Drive    1
3165 Upton Avenue           1
                           ..
4649 Worley Avenue          1
4155 Raccoon Run            1
1614 Heath

In [65]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [66]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [67]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [68]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [69]:
sum(treatments.auralin.isnull())

0

In [70]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [71]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
Import the cut treatments into a DataFrame and concatenate it with the original treatments DataFrame.

##### Code

In [72]:
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_clean = pd.concat([treatments_clean, treatments_cut],
                             ignore_index=True)

In [73]:
treatments_cut.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [74]:
treatments_cut.shape

(70, 7)

In [75]:
treatments_clean.shape

(350, 7)

##### Test

In [76]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [77]:
treatments_clean.tail()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
Recalculate the `hba1c_change` column: `hba1c_start` minus `hba1c_end`.

##### Code

In [78]:
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - treatments_clean.hba1c_end)

##### Test

In [79]:
treatments_clean.hba1c_change.head()

0    0.43
1    0.47
2    0.43
3    0.35
4    0.32
Name: hba1c_change, dtype: float64

### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

##### Code

In [80]:
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

##### Test

In [81]:
patients_clean['phone_number'].head(3)

0         951-719-9170
1    +1 (217) 569-3204
2         402-363-6804
Name: phone_number, dtype: object

In [82]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi       phone_number  
0  19.6       951-719-9170  
1  19.2  +1 (217) 569-3204  
2  24.8       402-363-6804  
3  31.7  +1 (732) 636-8246  
4  26.1       334-515-7487

In [83]:
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

In [84]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi       phone_number                      email  
0  19.6       951-719-9170   ZoeWellish@superrito.com  
1  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2  24.8       402-363-6804       JaeMDebord@gustr.com  
3  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4  26.1       334-515-7487        TimNeudorf@cuvox.de

In [85]:
patients_clean = patients_clean.drop('contact', axis=1)

In [86]:
patients_clean.head(1)

patient_id assigned_sex given_name  surname               address  \
0           1       female        Zoe  Wellish  576 Brown Bear Drive   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   

   height   bmi  phone_number                     email  
0      66  19.6  951-719-9170  ZoeWellish@superrito.com

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

##### Code

In [87]:
treatment_copy = treatments_clean.copy()

In [94]:
treatment_copy.head(3)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43

In [89]:
treatment_copy.shape

(350, 7)

In [135]:
treatments_clean = treatment_copy.copy()

In [136]:
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean.head(3)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1     elliot  richardson         7.56       7.09          0.47   auralin   
2   yukitaka    takenaka         7.68       7.25          0.43   auralin   

        dose  
0  41u - 48u  
1          -  
2          -

In [137]:
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean.head(3)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   

        dose  
0  41u - 48u  
3  33u - 36u  
6  37u - 42u

In [138]:
treatments_clean.tail(3)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
694      lixue       hsueh         9.21       8.80          0.41   novodra   
696      jakob    jakobsen         7.96       7.51          0.45   novodra   
698      berta  napolitani         7.68       7.21          0.47   novodra   

          dose  
694  22u - 23u  
696  28u - 26u  
698  42u - 44u

In [139]:
treatments_clean.shape

(350, 7)

In [140]:
pd.options.display.max_rows = 700
treatments_clean

given_name         surname  hba1c_start  hba1c_end  hba1c_change  \
0             veronika        jindrová         7.63       7.20          0.43   
3                 skye      gormanston         7.97       7.62          0.35   
6               sophia          haugen         7.65       7.27          0.38   
7                eddie          archer         7.89       7.55          0.34   
9                 asia         woźniak         7.76       7.37          0.39   
10              joseph             day         7.70       7.19          0.51   
12             roxanne       andreyeva         9.54       9.14          0.40   
14              simone    baumgaertner         7.74       7.30          0.44   
15                enco          žibrik         7.78       7.34          0.44   
16             camilla        zaitseva         7.53       7.13          0.40   
21               tekla         walczak         7.61       7.29          0.32   
22         brancaleone           russo         8.61       8.18          0.43   
24                isac            berg         9.68       9.29          0.39   
28             clinton          miller         7.79       7.40          0.39   
29              eugene         mironov         7.81       7.48          0.33   
30         szilveszter           totth         7.70       7.38          0.32   
34           alexander       mathiesen         7.96       7.55          0.41   
36                 chỉ             lâm         7.68       7.24          0.44   
37           władysław       wieczorek         7.92       7.47          0.45   
38            kristján         ingason         7.92       7.57          0.35   
42              marija        grubišić         7.53       7.15          0.38   
44               sauli      koivuniemi         7.67       7.37          0.30   
46             mariana           souza         7.86       7.51          0.35   
51          kristoffer       martinsen         9.18       8.64          0.54   
52                  mỹ           quynh         7.61       7.16          0.45   
54                oles         zhdanov         7.52       7.11          0.41   
55             triana.        terrazas         7.71       7.34          0.37   
58              gabryŝ     tomaszewski         7.87       7.47          0.40   
59           leixandre          alanis         7.74       7.32          0.42   
60        onyekachukwu          obinna         7.58       7.12          0.46   
61            mijndert       ouwendijk         7.81       7.50          0.31   
63             coralie         allaire         7.63       7.30          0.33   
64             cumelén      miramontes         7.71       7.33          0.38   
66              louise         johnson         7.63       7.32          0.31   
69                ivan           fomin         9.12       8.73          0.39   
73               mette       sandgreen         7.56       7.08          0.48   
78               julia          werner         7.58       7.24          0.34   
79               avdei        tikhonov         7.77       7.27          0.50   
81              robert          wagner         9.84       9.52          0.32   
82                maya        isaksson         7.66       7.17          0.49   
84              furuta           osman         7.52       7.18          0.34   
85              nilton     quintanilla         7.86       7.48          0.38   
87              finley        chandler         7.65       7.26          0.39   
89            stefanie          herman         7.72       7.30          0.42   
90               karen        jakobsen         7.59       7.25          0.34   
95              jaakko       eskelinen         7.73       7.42          0.31   
98              idalia           moore         8.34       7.90          0.44   
104              renzo        lucchese         7.54       7.17          0.37   
109             marius          hansen         7.97       7.53          0.44   
112         

In [142]:
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ',1).str

In [143]:
treatments_clean

given_name         surname  hba1c_start  hba1c_end  hba1c_change  \
0             veronika        jindrová         7.63       7.20          0.43   
3                 skye      gormanston         7.97       7.62          0.35   
6               sophia          haugen         7.65       7.27          0.38   
7                eddie          archer         7.89       7.55          0.34   
9                 asia         woźniak         7.76       7.37          0.39   
10              joseph             day         7.70       7.19          0.51   
12             roxanne       andreyeva         9.54       9.14          0.40   
14              simone    baumgaertner         7.74       7.30          0.44   
15                enco          žibrik         7.78       7.34          0.44   
16             camilla        zaitseva         7.53       7.13          0.40   
21               tekla         walczak         7.61       7.29          0.32   
22         brancaleone           russo         8.61       8.18          0.43   
24                isac            berg         9.68       9.29          0.39   
28             clinton          miller         7.79       7.40          0.39   
29              eugene         mironov         7.81       7.48          0.33   
30         szilveszter           totth         7.70       7.38          0.32   
34           alexander       mathiesen         7.96       7.55          0.41   
36                 chỉ             lâm         7.68       7.24          0.44   
37           władysław       wieczorek         7.92       7.47          0.45   
38            kristján         ingason         7.92       7.57          0.35   
42              marija        grubišić         7.53       7.15          0.38   
44               sauli      koivuniemi         7.67       7.37          0.30   
46             mariana           souza         7.86       7.51          0.35   
51          kristoffer       martinsen         9.18       8.64          0.54   
52                  mỹ           quynh         7.61       7.16          0.45   
54                oles         zhdanov         7.52       7.11          0.41   
55             triana.        terrazas         7.71       7.34          0.37   
58              gabryŝ     tomaszewski         7.87       7.47          0.40   
59           leixandre          alanis         7.74       7.32          0.42   
60        onyekachukwu          obinna         7.58       7.12          0.46   
61            mijndert       ouwendijk         7.81       7.50          0.31   
63             coralie         allaire         7.63       7.30          0.33   
64             cumelén      miramontes         7.71       7.33          0.38   
66              louise         johnson         7.63       7.32          0.31   
69                ivan           fomin         9.12       8.73          0.39   
73               mette       sandgreen         7.56       7.08          0.48   
78               julia          werner         7.58       7.24          0.34   
79               avdei        tikhonov         7.77       7.27          0.50   
81              robert          wagner         9.84       9.52          0.32   
82                maya        isaksson         7.66       7.17          0.49   
84              furuta           osman         7.52       7.18          0.34   
85              nilton     quintanilla         7.86       7.48          0.38   
87              finley        chandler         7.65       7.26          0.39   
89            stefanie          herman         7.72       7.30          0.42   
90               karen        jakobsen         7.59       7.25          0.34   
95              jaakko       eskelinen         7.73       7.42          0.31   
98              idalia           moore         8.34       7.90          0.44   
104              renzo        lucchese         7.54       7.17          0.37   
109             marius          hansen         7.97       7.53          0.44   
112         

In [109]:
treatments_clean = treatments_clean.drop('dose', axis=1)

In [110]:
treatments_clean

given_name         surname  hba1c_start  hba1c_end  hba1c_change  \
0             veronika        jindrová         7.63       7.20          0.43   
3                 skye      gormanston         7.97       7.62          0.35   
6               sophia          haugen         7.65       7.27          0.38   
7                eddie          archer         7.89       7.55          0.34   
9                 asia         woźniak         7.76       7.37          0.39   
10              joseph             day         7.70       7.19          0.51   
12             roxanne       andreyeva         9.54       9.14          0.40   
14              simone    baumgaertner         7.74       7.30          0.44   
15                enco          žibrik         7.78       7.34          0.44   
16             camilla        zaitseva         7.53       7.13          0.40   
21               tekla         walczak         7.61       7.29          0.32   
22         brancaleone           russo         8.61       8.18          0.43   
24                isac            berg         9.68       9.29          0.39   
28             clinton          miller         7.79       7.40          0.39   
29              eugene         mironov         7.81       7.48          0.33   
30         szilveszter           totth         7.70       7.38          0.32   
34           alexander       mathiesen         7.96       7.55          0.41   
36                 chỉ             lâm         7.68       7.24          0.44   
37           władysław       wieczorek         7.92       7.47          0.45   
38            kristján         ingason         7.92       7.57          0.35   
42              marija        grubišić         7.53       7.15          0.38   
44               sauli      koivuniemi         7.67       7.37          0.30   
46             mariana           souza         7.86       7.51          0.35   
51          kristoffer       martinsen         9.18       8.64          0.54   
52                  mỹ           quynh         7.61       7.16          0.45   
54                oles         zhdanov         7.52       7.11          0.41   
55             triana.        terrazas         7.71       7.34          0.37   
58              gabryŝ     tomaszewski         7.87       7.47          0.40   
59           leixandre          alanis         7.74       7.32          0.42   
60        onyekachukwu          obinna         7.58       7.12          0.46   
61            mijndert       ouwendijk         7.81       7.50          0.31   
63             coralie         allaire         7.63       7.30          0.33   
64             cumelén      miramontes         7.71       7.33          0.38   
66              louise         johnson         7.63       7.32          0.31   
69                ivan           fomin         9.12       8.73          0.39   
73               mette       sandgreen         7.56       7.08          0.48   
78               julia          werner         7.58       7.24          0.34   
79               avdei        tikhonov         7.77       7.27          0.50   
81              robert          wagner         9.84       9.52          0.32   
82                maya        isaksson         7.66       7.17          0.49   
84              furuta           osman         7.52       7.18          0.34   
85              nilton     quintanilla         7.86       7.48          0.38   
87              finley        chandler         7.65       7.26          0.39   
89            stefanie          herman         7.72       7.30          0.42   
90               karen        jakobsen         7.59       7.25          0.34   
95              jaakko       eskelinen         7.73       7.42          0.31   
98              idalia           moore         8.34       7.90          0.44   
104              renzo        lucchese         7.54       7.17          0.37   
109             marius          hansen         7.97       7.53          0.44   
112         

In [90]:
# Your cleaning code here
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
                           
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str
treatments_clean = treatments_clean.drop('dose', axis=1)

##### Test

In [93]:
# Your testing code here
treatments_clean.head(3)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u

In [92]:
treatments_clean.shape

(350, 8)

#### Adverse reaction should be part of the `treatments` table

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
*Your definition here. Hint: see the "Data Cleaning Process" page.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Dsvid Gustafsson

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Multiple phone number formats

##### Define
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Default John Doe data

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### kgs instead of lbs for Zaitseva weight

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here